marlowe-spec-test/src/Marlowe/Spec/Core/Arbitrary.hs

```diff
@@ -365,16 +503,226 @@ genTransactionOutput :: InterpretJsonRequest -> GenT IO TransactionOutput
 genTransactionOutput i =
  frequency
     [ (30, TransactionError <$> liftGen genTransactionError)
-    , (70, do
-              wSize <- liftGen $ chooseInt (0, 2)
-              warnings <- vectorOf wSize $ genTransactionWarning i
-              pSize <- liftGen $ chooseInt (0, 2)
-              payments <- vectorOf pSize $ genPayment i
-              state <- resize 2 $ genState i
-              contract <- resize 2 $ genContract i
-              pure $ TransactionOutput $ TransactionOutputRecord_ext warnings payments state contract ()
-      )
+    , (70, TransactionOutput <$> do
+                                   wSize <- liftGen $ chooseInt (0, 2)
+                                   warnings <- vectorOf wSize $ genTransactionWarning i
+                                   pSize <- liftGen $ chooseInt (0, 2)
+                                   payments <- vectorOf pSize $ genPayment i
+                                   state <- resize 2 $ genState i
+                                   contract <- resize 2 $ genContract i
+                                   pure $ TransactionOutputRecord_ext warnings payments state contract ())
     ]
 
 genEnvironment :: Gen (Environment_ext ())
 genEnvironment = Environment_ext <$> genInterval <*> pure ()
+
+-- | Generate a random step for a contract.
+arbitraryValidStep :: State_ext ()             -- ^ The state of the contract.
+                   -> Contract                 -- ^ The contract.
+                   -> Gen (Transaction_ext ()) -- ^ Generator for a transaction input for a single step.
+arbitraryValidStep _ (When [] timeout _) = Transaction_ext <$> arbitraryTimeIntervalAfter timeout <*> pure [] <*> pure ()
+arbitraryValidStep state (When cases timeout _) =
```

Comment:
If I understood this function correctly, the name might be misleading. It seems to recurse the arbitrary inputs until it reaches to a close, so the meaning is actually captured by `arbitraryValidInputs` defined below.